### PR TITLE
Animations panel

### PR DESCRIPTION
**Pull request by emackey**
Cesium previews now have a small UI for turning individual animations on and off.

Note that many of our sample models have a known issue where each animation drives a single channel, rather than each animation corresponding to a different cycle.  So for the BrainStem sample, for example, it only makes sense to turn all animations on and off at the same time, because they should all be a single animation.  There are already issues open on respective repos to correct this problem.

### CODE REVIEW

**Comment by HowardWolosky** [comment]
Haven't really reviewed the code in detail yet, but I took a quick glimpse at the implementation and then tried it out myself.

Initial thoughts:

1. I like the idea.  Thanks for implementing it.  I can definitely see there being value in being able to dynamically start/stop animations within the preview.

2. Might be better to leverage the `name` attribute from the animation (if it exists).  For instance, the [Animated Morph Sphere](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/AnimatedMorphSphere/glTF/AnimatedMorphSphere.glTF) named its only animation as "Globe".

3. It might be interesting to unify the "controls" all into the existing common control menu that we're using.  You already have the concept of the content in there changing with your dynamically changing checkbox for showing the background.  After `previewModel.js` replaces the engine content in `updatePreview`, it could call a method that we'd expect the preview engines to expose (similar to how we already call `cleanup()`), that would return back an array of animations that can be enabled/disabled.  It would then call a new method (`toggleAnimation(int)`) that the engines would expose in order to do the animations.  We could then add similar support for Babylon and three in a subsequent check-in, without having to duplicate all the UI work in each of those preview engines as well.  I realize that this approach would undo some of the work that you've already done, but it feels like the more maintainable/sustainable/user-friendly approach to the design.  On the other hand, maybe an argument could be made that each engine might want/need its own custom menu to control engine-specific stuff.  At the moment though, this doesn't feel engine-specific.

**Comment by emackey** [comment]
Hey @HowardWolosky, thanks for reviewing this!

On the topic of unifying controls, how would you feel if I switched the shared UI over to knockout-based instead of DAT GUI?  The animations panel here is an example of a Knockout data-bound UI.  The HTML gets extra `data-bind` attributes that cause it to bind to a view model maintained in JS.  This makes it much cleaner than DAT to add and remove entire custom sections of the UI.  The engines would share a parent viewModel, and each engine could possibly add an engine-specific sub-view-model, or enable and disable shared sections that only apply to certain engines (like the background).

**Comment by HowardWolosky** [comment]
That sounds like a great idea.  I'm not attached to DAT GUI at all...it just happened to be what I was familiar with since it was already being used in some other projects that I had dabbled in.  From your description, Knockout sounds like it's a much better fit for the needs of this project.

**Comment by emackey** [comment]
Beware the September release of VSCode 1.16.0.  They changed [the way extension debugging works](https://code.visualstudio.com/updates/v1_16#_extension-authoring) and I don't have it working yet.

**Comment by emackey** [comment]
Got this working again, by un-installing the 32-bit version and installing the 64-bit version of 1.16.0.  Going to take a bit longer to move the UIs around due to some other priorities.  Maybe I'll find time this weekend.

**Comment by emackey** [comment]
Well... That got out-of-hand fairly quickly.

**Comment by emackey** [comment]
Actually I'm going to tweak the styling just a bit more and then this will be ready.

Fixes #12 (or converts it to OBE really.  Too lazy to warn the user that the model has loaded from disk, since in practice I find myself using the saved version a lot more than the live edits.  I can fix this if it's objectionable, I would just need to build a knockout panel for it).

**Comment by emackey** [comment]
This is ready again.

Apologies if I took a bunch of familiar code and made it alien.  Basically, all the data that is exposed on the UI now lives in a structure called `mainViewModel`.  Each primitive element of data is wrapped in a Knockout "observable", which provides that piece of data a `subscribe` function, such that interested parties may be notified of any changes to the data.  It makes the getters and setters look strange: To get a value, you must put empty parentheses at the end (`var x = myValue();`), and to set the value, you don't use an equal sign, you pass the new value into it (`myValue(x);`).  Actually there's a way to avoid this and get standard get/set behavior back, but it makes subscribing to things more difficult, so it's a tradeoff.

Anyway, the UI is data-bound to this viewModel, mostly via `data-bind` attributes in the HTML document itself.  This is a 2-way binding, such that user interactions on the UI are applied to the data as you expect, but also, external programmatic changes to the data are propagated back to the UI.  Engine code that drives things such as starting animations or enabling the background can just use 1-way binding, subscribing to the observable Boolean value that controls the particular thing, and implementing the change when the subscription fires.  More complex actions like the "all" and "none" buttons are handled simply, by iterating through the available actions and assigning `true` or `false` to their `active` flags, trusting that both the UI and the rendering engine will respond to such changes, and trusting Knockout to only fire subscriptions on actual changes (not on re-assignments of the same value).

Thanks for taking the time to review this!

**Comment by emackey** [comment]
@HowardWolosky Are you still planning/willing to look at this?

**Comment by HowardWolosky** [comment]
Indeed. (Especially after convincing you to do all that extra work). Just not until Mon or Tues though due to time off.  If you don't want to wait though, you can always apply any feedback in a subsequent check-in...your call.

**Comment by emackey** [comment]
Awesome.  Enjoy your time off!

**Comment by HowardWolosky** [comment]
I just tried it out in the Extension Host and your change looks awesome.  I think unifying the controls to a single menu area definitely gives a better end-user experience.

Reviewing the code now.

File: pages/previewModel.html [hunk 1]
```diff
@@ -1,76 +1,51 @@
...
+                <span class="angleBox"><span></span></span>
+                <span>Show Background</span>
+            </label>
+        </div>
+        <div class="rowUI animationUI" style="display: none;" data-bind="visible: animations().length > 0">
+            <div>animations:</div>
+            <div class="animationList" data-bind="foreach: animations">
+                <label class="checkboxLabel" data-bind="css: { active: active() }">
+                    <input type="checkbox" data-bind="checked: active">
+                    <span class="angleBox"><span></span></span>
```
[the remark this answers is not in the document]
Comment (emackey) [reply pages/previewModel.html]: Yeah, actually that's a `span` within a `span.angleBox`, with two open and two close tags.  This is just for the sake of drawing goofy angled edges with CSS.  With a bit more effort I could have made a custom knockout binding to a single span that hides the detail of the inner span and the label text binding.  Or, you know, I could have let the browser just put a normal checkbox there.

File: src/gltfPreviewDocumentContentProvider.ts [hunk 2]
```diff
@@ -69,86 +77,47 @@ export class GltfPreviewDocumentContentProvider implements TextDocumentContentPr
...
+            'engines/Three/three.min.js',
+            'engines/Three/GLTF2Loader.js',
+            'engines/Three/OrbitControls.js',
+            'pages/babylonView.js',
+            'pages/cesiumView.js',
+            'pages/threeView.js',
+            'pages/previewModel.js'
+        ];
+
+        const content = this._mainHtml.replace('{assets}',
```
[the remark this answers is not in the document]
Comment (emackey) [reply src/gltfPreviewDocumentContentProvider.ts]: I figure it leaves the door open to replace `window.ko` with a real copy of Knockout if we ever have the need to separate that out.  But for now, it's bundled with Cesium, so no need to include it twice.

File: pages/previewModel.js [hunk 3]
```diff
@@ -1,103 +1,73 @@
...
+    html: 'cesiumHtml',
+    view: CesiumView
+}, {
+    name: 'Three.js',
+    html: 'threeHtml',
+    view: ThreeView
+}];
+
+// Use Cesium's built-in copy of Knockout as the global UI manager
+window.ko = Cesium.knockout;
```
[the remark this answers is not in the document]
Comment (emackey) [reply pages/previewModel.js]: I figure it leaves the door open to replace `window.ko` with a real copy of Knockout if we ever have the need to separate that out.  But for now, it's bundled with Cesium, so no need to include it twice.

File: src/gltfPreviewDocumentContentProvider.ts [hunk 4]
```diff
@@ -69,86 +77,47 @@ export class GltfPreviewDocumentContentProvider implements TextDocumentContentPr
...
+            styles.map(s => `<link rel="stylesheet" href="${this.getFilePath(s)}"></link>\n`).join('') +
+            strings.map(s => `<script id="${s.id}" type="text/plain">${s.text}</script>\n`).join('') +
+            scripts.map(s => `<script type="text/javascript" src="${this.getFilePath(s)}"></script>\n`).join(''));
```
[the remark this answers is not in the document]
Comment (emackey) [reply src/gltfPreviewDocumentContentProvider.ts]: Yeah, but I suspect the main memory hog is the WebGL canvases and runtimes behind them.  These little viewer scripts are just a couple KB each, no sweat having all 3 in memory at once, so long as all 3 WebGL engines aren't all running at once.

When you switch tabs in this new version, the `cleanup` routine is called on the old engine, and my hope is this releases as much memory as possible.  Certainly the old WebGL canvas is removed from the DOM.  I don't recall if I made sure that the old render loops are all halted, that would be a performance problem if they aren't.  The Cesium one probably needs a `viewer.destroy()` or something on it if it doesn't have one already.

[the remark this answers is not in the document]
Comment (emackey) [reply src/gltfPreviewDocumentContentProvider.ts]: Oh right there's no `viewer`, just a `scene`.  Probably still needs a `scene.destroy()`.  The render loop for Cesium at least does appear to halt when `enabled` goes `false`.

**Pull request merged**